crates/contract/src/lib.rs

```diff
@@ -1322,6 +1325,40 @@ impl MpcContract {
             Err(err) => env::panic_str(&format!("not a voter, {:?}", err)),
         }
     }
+    /// Ensures that the caller is an attested participant
+    /// in the currently active protocol phase.
+    ///
+    /// Active phases:
+    /// - `Initializing` → uses proposed participants from generating_key
+    /// - `Running` → uses current active participants
+    /// - `Resharing` → uses new participants from resharing proposal
+    ///
+    /// Panics if:
+    /// - The protocol is not active (e.g., NotInitialized)
+    /// - The caller is not attested or not in the relevant participants set
+    pub fn assert_caller_is_attested_participant_and_protocol_active(&self) {
+        // Select participants based on the current protocol state
+        let participants = match &self.protocol_state {
+            ProtocolContractState::Initializing(state) => {
+                state.generating_key.proposed_parameters().participants()
+            }
+            ProtocolContractState::Running(state) => state.parameters.participants(),
+            ProtocolContractState::Resharing(state) => {
+                state.resharing_key.proposed_parameters().participants()
+            }
+            ProtocolContractState::NotInitialized => {
+                panic!("Protocol must be Initializing, Running, or Resharing to perform this operation");
+            }
+        };
```

Comment:
nit: might be nice to make add this to `state.rs` and make it a method of `ProtocolContracteState`.
https://github.com/near/mpc/blob/3d83f1f4867d2aa43eb274df2d809419e690a004/crates/contract/src/state.rs#L179